tests/end_to_end/end_to_end.cpp

```diff
@@ -829,13 +834,181 @@ map_test(ebpf_execution_type_t execution_type)
     bpf_object__close(object);
 }
 
+bpf_sock_ops_t
+populate_bpf_sock_ops(
+    bpf_sock_op_type_t op,
+    uint8_t address_family,
+    const std::string& source_address,
+    uint16_t source_port,
+    const std::string& destination_address,
+    uint16_t destination_port,
+    uint8_t ip_protocol)
+{
+    bpf_sock_ops_t return_value{op, address_family};
+    const char* terminator;
+    if (address_family == AF_INET) {
+        in_addr source_ip_address;
+        in_addr destination_ip_address;
+        REQUIRE(RtlIpv4StringToAddressA(source_address.c_str(), false, &terminator, &source_ip_address) == 0);
+        REQUIRE(RtlIpv4StringToAddressA(destination_address.c_str(), false, &terminator, &destination_ip_address) == 0);
+        return_value.local_ip4 = source_ip_address.S_un.S_addr;
+        return_value.remote_ip4 = destination_ip_address.S_un.S_addr;
+    } else if (address_family == AF_INET6) {
+        in_addr6 source_ip_address;
+        in_addr6 destination_ip_address;
+        REQUIRE(RtlIpv6StringToAddressA(source_address.c_str(), &terminator, &source_ip_address) == 0);
+        REQUIRE(RtlIpv6StringToAddressA(destination_address.c_str(), &terminator, &destination_ip_address) == 0);
+        memcpy(return_value.local_ip6, source_ip_address.u.Byte, sizeof(return_value.local_ip6));
+        memcpy(return_value.remote_ip6, destination_ip_address.u.Byte, sizeof(return_value.remote_ip6));
+    }
+    return_value.local_port = htons(source_port);
+    return_value.remote_port = htons(destination_port);
+    return_value.protocol = ip_protocol;
+    return return_value;
+}
+
+typedef struct _ip_address
+{
+    union
+    {
+        uint32_t ipv4;
+        uint32_t ipv6[4];
+    };
+} ip_address_t;
+
+typedef struct _connection_tuple
+{
+    ip_address_t src_ip;
+    uint16_t src_port;
+    ip_address_t dst_ip;
+    uint16_t dst_port;
+    uint32_t protocol;
+} connection_tuple_t;
+
+typedef struct _connection_history
+{
+    connection_tuple_t tuple;
+    bool ipv4;
+    uint64_t start_time;
+    uint64_t end_time;
+} connection_history_t;
+
+std::string
+ip_address_to_string(bool ipv4, const ip_address_t& ip_address)
+{
+    std::string buffer;
+    if (ipv4) {
+        buffer.resize(16);
+        in_addr addr;
+        addr.S_un.S_addr = ip_address.ipv4;
+        RtlIpv4AddressToStringA(&addr, buffer.data());
+
+    } else {
+        buffer.resize(46);
+        in_addr6 addr;
+        memcpy(addr.u.Byte, ip_address.ipv6, sizeof(ip_address.ipv6));
+        RtlIpv6AddressToStringA(&addr, buffer.data());
+        buffer = "[" + buffer + "]";
+    }
+    return buffer;
+}
+
+void
+conn_track_test(ebpf_execution_type_t execution_type)
```

Comment:
coding style is to avoid abbreviations in code
```suggestion
connection_track_test(ebpf_execution_type_t execution_type)
```